Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc

```diff
@@ -0,0 +1,177 @@
+                                                                    iConfig.getParameter<std::string>("detectorTag")))},
+      gdRealTokens_{setWhatProduced(this, &PPSGeometryESProducer::produceRealGD)},
+      gdMisTokens_{setWhatProduced(this, &PPSGeometryESProducer::produceMisalignedGD)},
+      dgdRealToken_{
+          setWhatProduced(this, &PPSGeometryESProducer::produceRealTG).consumes<DetGeomDesc>(edm::ESInputTag())},
+      dgdMisToken_{
+          setWhatProduced(this, &PPSGeometryESProducer::produceMisalignedTG).consumes<DetGeomDesc>(edm::ESInputTag())} {
+}
+
+void PPSGeometryESProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.addUntracked<unsigned int>("verbosity", 1);
+  desc.add<std::string>("detectorTag", std::string());
+  descriptions.add("PPSGeometryESProducer", desc);
+}
+
+//----------------------------------------------------------------------------------------------------
+/*
+ * Apply alignments by doing a BFS on idealGD tree.
+ */
+void PPSGeometryESProducer::applyAlignments(const DetGeomDesc& idealGD,
+                                            const CTPPSRPAlignmentCorrectionsData* alignments,
+                                            DetGeomDesc*& newGD) {
+  newGD = new DetGeomDesc(idealGD);
+  std::deque<const DetGeomDesc*> buffer;
+  std::deque<DetGeomDesc*> bufferNew;
+  buffer.emplace_back(&idealGD);
+  bufferNew.emplace_back(newGD);
+
+  while (!buffer.empty()) {
+    const DetGeomDesc* sD = buffer.front();
+    DetGeomDesc* pD = bufferNew.front();
+    buffer.pop_front();
+    bufferNew.pop_front();
+
+    const std::string name = pD->name();
+
+    // Is it sensor? If yes, apply full sensor alignments
+    if (name == DDD_TOTEM_RP_SENSOR_NAME || name == DDD_CTPPS_DIAMONDS_SEGMENT_NAME ||
+        name == DDD_CTPPS_UFSD_SEGMENT_NAME || name == DDD_CTPPS_PIXELS_SENSOR_NAME ||
+        std::regex_match(name, std::regex(DDD_TOTEM_TIMING_SENSOR_TMPL))) {
+      unsigned int plId = pD->geographicalID();
+
+      if (alignments) {
+        const auto& ac = alignments->getFullSensorCorrection(plId);
+        pD->applyAlignment(ac);
+      }
+    }
+
+    // Is it RP box? If yes, apply RP alignments
+    if (name == DDD_TOTEM_RP_RP_NAME || name == DDD_CTPPS_DIAMONDS_RP_NAME || name == DDD_CTPPS_PIXELS_RP_NAME ||
+        name == DDD_TOTEM_TIMING_RP_NAME) {
+      unsigned int rpId = pD->geographicalID();
+
+      if (alignments) {
+        const auto& ac = alignments->getRPCorrection(rpId);
+        pD->applyAlignment(ac);
+      }
+    }
+
+    // create and add children
+    for (unsigned int i = 0; i < sD->components().size(); i++) {
+      const DetGeomDesc* sDC = sD->components()[i];
+      buffer.emplace_back(sDC);
+
+      // create new node with the same information as in sDC and add it as a child of pD
+      DetGeomDesc* cD = new DetGeomDesc(*sDC);
+      pD->addComponent(cD);
+
+      bufferNew.emplace_back(cD);
```

Comment:
same here, I'm not sure where `cD` gets destroyed

Reply:
This is the exact code from 'old DD' Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc
This code was not rewritten, and re-writing PPS code would be a significant, independent effort (in which case there would be many more points to clean / modernize). 
This was not the topic of this PR, and I think should not be in this PR, as I think a PR should be one identifiable action. Here the action is to fully move CTPPS to DD4hep.
I think if someone would like to clean / modernize PPS code, that would deserve another, independent PR.

Reply:
Instead here, I think it would be essential to cross-check / review that there is absolutely no unexpected regression from old DD -> DD4hep / nor any accidental remaining from old DD.